tests/bluetooth/bsim_bt/bsim_test_mesh_pst/src/settings_test_backend.c

```diff
@@ -0,0 +1,291 @@
+		return 0;
+	}
+
+	char val[ENTRY_VAL_MAX_LEN];
+	int br = read(fp, val, len * 2);
+	if (br < len * 2) {
+		return 0;
+	}
+
+	uint8_t n;
+	for (int i = 0; i < len; i++) {
+		sscanf(&val[i * 2], "%2hhx", &n);
+		((uint8_t *)data)[i] = n;
+	}
+
+	return len;
+}
+
+static int settings_custom_load(struct settings_store *cs, const struct settings_load_arg *arg)
+{
+	char name[ENTRY_NAME_MAX_LEN];
+	int namelen;
+	off_t curoff = 0;
+	struct line_read_ctx valctx;
+
+	do {
+		namelen = get_next_entry_name(curoff, &valctx.offset,
+					      name, sizeof(name), &valctx.len);
+
+		if (namelen <= 0) {
+			break;
+		}
+
+		LOG_INF("name: %s", name);
+
+		/* check for matching subtree */
+		if (arg->subtree != NULL && !strstr(name, arg->subtree)) {
+			continue;
+		}
+
+		int err = settings_call_set_handler(name, namelen, settings_line_read_cb,
+						    &valctx, arg);
+
+		if (err < 0) {
+			return err;
+		}
+
+		curoff = lseek(fp, 0, SEEK_CUR);
+	} while (curoff != EOF);
+
+	return 0;
+}
+
+/* Entries are saved to optimize redability of the settings file for test development and
+ * debugging purposes. Format:
+ * <length-hex-str>,<entry-key>,<entry-value-hex-str>
+ */
+static int settings_custom_save(struct settings_store *cs, const char *name,
+				const char *value, size_t val_len)
+{
+	if (strlen(name) > ENTRY_NAME_MAX_LEN || val_len > ENTRY_VAL_MAX_LEN) {
+		return -1;
+	}
+
+	char bufname[ENTRY_NAME_MAX_LEN + ENTRY_LEN_SIZE + sizeof(ENTRY_SEPERATOR) * 2];
+	char bufval[ENTRY_VAL_MAX_LEN * 2 + 1];
+	size_t namecnt = snprintk(bufname, sizeof(bufname), "%04x%c%s%c",
+				  val_len * 2, ENTRY_SEPERATOR, name, ENTRY_SEPERATOR);
+	size_t valcnt = 0;
+
+	int offset = entry_check_and_seek(name);
+
+	if (offset == 0) {
+		if (lseek(fp, 0, SEEK_END) < 0 ||
+		    write(fp, bufname, namecnt) < namecnt) {
+			return -1;
+		}
+	}
+
+	while (valcnt < (val_len * 2)) {
+		valcnt += snprintk(&bufval[valcnt], 3, "%02x",
+				   (uint8_t)value[valcnt / 2]);
+	};
+
+	/* helps in making settings file redable */
+	bufval[valcnt++] = '\n';
+
+	if (write(fp, bufval, valcnt) < valcnt) {
+		return -1;
+	}
+
+	return 0;
+}
+
+/* custom backend interface */
+static struct settings_store_itf settings_custom_itf = {
+	.csi_load = settings_custom_load,
+	.csi_save = settings_custom_save,
+};
+
+/* custom backend node */
+static struct settings_store settings_custom_store = {
+	.cs_itf = &settings_custom_itf
+};
+
+int settings_backend_init(void)
+{
+	if (!backed_init) {
+		LOG_ERR("Call settings_test_backend_preinit() before starting test");
+		return -1;
+	}
+
+	/* register custom backend */
+	settings_dst_register(&settings_custom_store);
+	settings_src_register(&settings_custom_store);
+	return 0;
+}
+
+int settings_test_backend_preinit(enum test_file_state fmode)
+{
+	if (getcwd(backend_file, sizeof(backend_file)) == NULL) {
+		return -1;
+	}
+
+	strcat(backend_file, SETTINGS_FILE);
+
+	LOG_INF("file path: %s", backend_file);
+
+	if (fmode == FILE_CLEAR) {
```

Comment:
Probably, I just started using raw POSIX APIs and continued with it. This seems to be working without any issue.

Reply:
I agree with @alxelax, this is would be cleaner and easier to maintain if it used stdio. 

Additionally, the settings entries are constructed to map to a filesystem, with its path-like naming scheme. This could be utilized here as well by storing each settings entry in a separate file named the same as its settings entry. Either use mkdir to create an actual directory tree, or replace the slashes with `$` or something to create a flat representation. In the end, all we'd need is `fopen()`, `fgets()` and `fputs()`, plus `mkdir()`, if you feel like it.

Reply:
Mapping can be done in several ways, however, it will be over-complication to map the settings tree to individual files. I found that having all entries in a single file makes it much easier to debug the settings subsystem or tests. I will prefer to use a single-file approach.